File: backend/shared/logo_url_policy.py
"""Politique unique des URLs de logo persistées (company / institution).

Contrat :
- persisté : chemin ``/uploads/...`` ou URL ``https://``
- refusé : http, data, blob, javascript, file, vbscript, protocol-relative
- ``blob:`` n'est jamais une valeur persistée
"""

from __future__ import annotations

import re
from urllib.parse import urlparse

LOGO_URL_DENIED_MESSAGE = "URL de logo non autorisée"
MAX_LOGO_URL_LENGTH = 500
_UPLOADS_PATH_RE = re.compile(r"^/uploads/[A-Za-z0-9._/-]+$")


class InvalidLogoUrl(ValueError):
    """URL de logo hors politique."""


def normalize_persisted_logo_url(raw: str | None) -> str | None:
    """Valide et normalise une URL de logo destinée à la base.

    Returns:
        None si vide, sinon le chemin ``/uploads/...`` ou une URL https reconstruite.

    Raises:
        InvalidLogoUrl: si la valeur n'est pas autorisée.
    """
    if raw is None:
        return None
    if not isinstance(raw, str):
        raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)

    value = raw.strip()
    if not value:
        return None
    if len(value) > MAX_LOGO_URL_LENGTH:
        raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)
    if any(ch in value for ch in ("\r", "\n", "\x00")):
        raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)
    if value.startswith("//"):
        raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)

    if value.startswith("/"):
        path = value.split("?", 1)[0].split("#", 1)[0]
        if ".." in path or "//" in path:
            raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)
        if not _UPLOADS_PATH_RE.fullmatch(path):
            raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)
        return path

    parsed = urlparse(value)
    scheme = (parsed.scheme or "").lower()
    if scheme != "https":
        raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)
    if parsed.username or parsed.password:
        raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)
    hostname = parsed.hostname
    if not hostname:
        raise InvalidLogoUrl(LOGO_URL_DENIED_MESSAGE)

    host = hostname
    if parsed.port:
        host = f"{hostname}:{parsed.port}"
    path = parsed.path or "/"
    query = f"?{parsed.query}" if parsed.query else ""
    return f"https://{host}{path}{query}"

File: frontend/src/utils/logoUrlPolicy.js
/**
 * Politique unique des URLs de logo (rendu <img> + persistance).
 *
 * Persisté : /uploads/... ou https://
 * Preview locale : blob: uniquement si allowPreview
 * Refusé : javascript, data, file, vbscript, http utilisateur, //protocol-relative
 */

export const LOGO_URL_DENIED_MESSAGE = 'URL de logo non autorisée';
export const MAX_LOGO_URL_LENGTH = 500;

const UPLOADS_PATH_RE = /^\/uploads\/[A-Za-z0-9._/-]+$/;

function fail() {
  return { ok: false, value: null };
}

function isSafeUploadsPath(path) {
  if (!path || path.includes('..') || path.includes('//')) {
    return false;
  }
  return UPLOADS_PATH_RE.test(path);
}

/**
 * Valide une URL destinée à être stockée (jamais blob:/data:).
 * @param {unknown} raw
 * @returns {{ ok: boolean, value: string | null }}
 */
export function validatePersistedLogoUrl(raw) {
  if (raw == null) {
    return { ok: true, value: null };
  }
  if (typeof raw !== 'string') {
    return fail();
  }
  const value = raw.trim();
  if (!value) {
    return { ok: true, value: null };
  }
  if (value.length > MAX_LOGO_URL_LENGTH) {
    return fail();
  }
  if (/[\r\n\0]/.test(value)) {
    return fail();
  }
  if (value.startsWith('//')) {
    return fail();
  }

  if (value.startsWith('/')) {
    const path = value.split('?')[0].split('#')[0];
    if (!isSafeUploadsPath(path)) {
      return fail();
    }
    return { ok: true, value: path };
  }

  let parsed;
  try {
    parsed = new URL(value);
  } catch {
    return fail();
  }
  if (parsed.protocol !== 'https:') {
    return fail();
  }
  if (parsed.username || parsed.password) {
    return fail();
  }
  if (!parsed.hostname) {
    return fail();
  }
  return {
    ok: true,
    value: `https://${parsed.host}${parsed.pathname}${parsed.search}`,
  };
}

/**
 * Autorise un blob: local pour la prévisualisation avant upload.
 * @param {unknown} raw
 * @returns {string}
 */
export function resolvePreviewBlobUrl(raw) {
  if (typeof raw !== 'string') {
    return '';
  }
  const value = raw.trim();
  if (!value.toLowerCase().startsWith('blob:')) {
    return '';
  }
  try {
    const parsed = new URL(value);
    if (parsed.protocol !== 'blob:') {
      return '';
    }
    return parsed.href;
  } catch {
    return '';
  }
}
